README.md
# CodeBuddy

CodeBuddy is an AI-powered developer assistant for code reviews, automating commits, and more. This CLI tool is designed to streamline your development process and make your life easier.

[![npm version](https://badge.fury.io/js/%40joelybahh%2Fcode-buddy.svg)](https://www.npmjs.com/package/@joelybahh/code-buddy)
[![Build Status](https://travis-ci.com/joelybahh/code-buddy.svg?branch=main)](https://travis-ci.com/joelybahh/code-buddy)
[![Coverage Status](https://coveralls.io/repos/github/joelybahh/code-buddy/badge.svg?branch=main)](https://coveralls.io/github/joelybahh/code-buddy?branch=main)
[![License: ISC](https://img.shields.io/badge/License-ISC-blue.svg)](https://opensource.org/licenses/ISC)

## Features

-   Automatically generate commit messages based on the diff
-   AI-powered assistance for code reviews [Coming Soon]

## Installation

CodeBuddy can be installed globally using npm, Or you can view it on the npm registry [here](https://www.npmjs.com/package/@joelybahh/code-buddy).

```
npm install -g @joelybahh/code-buddy
```

Or using yarn:

```
yarn global add @joelybahh/code-buddy
```

## Configuration

CodeBuddy can be configured using a `cb.config.js` file in the root of your project. The following properties are available:

```
{
    chatGPT: {
        apiKey: string;
        organization: string;
        model: "gpt-4" | "gpt-3.5-turbo";
        maxTokens?: number;
        temperature?: number;
        topP?: number;
        presencePenalty?: number;
        frequencyPenalty?: number;
        stop?: string[];
    };
    commit?: {
        scopeTrim: string;
        issue?: {
            detectKey?: boolean;
            keyRegex?: RegExp;
            fallbackKey?: string;
        };
        format?: {
            sentenceCase?: boolean;
        };
        scope?: {
            mode?: "monorepo" | "traditional";
            srcDir?: string;
        }
    };
    diff?: {
        maxSize?: number;
        exclude?: string[];
    };
}
```

### chatGPT

This object contains properties related to ChatGPT authentication and requests.

-   `apiKey` (string): Your OpenAI API key for authentication.
-   `organization` (string): The organization ID associated with your OpenAI account.
-   `model` (string): The GPT model to use, either `"gpt-4"` or `"gpt-3.5-turbo"`.
-   `maxTokens` (number, optional): The maximum number of tokens to generate. Defaults to `200`.
-   `temperature` (number, optional): The temperature to use for token generation. Defaults to `0.3`.
-   `topP` (number, optional): The topP value to use for token generation. Defaults to `1`.
-   `presencePenalty` (number, optional): The presence penalty to use for token generation. Defaults to `0.5`.
-   `frequencyPenalty` (number, optional): The frequency penalty to use for token generation. Defaults to `0.5`.
-   `stop` (string[], optional): An array of strings to use as stop sequences for token generation.

### commit (optional)

This object contains properties related to commit message generation and formatting.

-   `scopeTrim` (string): A string to automatically trim from the scope of the commit message.
-   `issue` (optional): An object containing properties related to issue key handling in commit messages.
    -   `detectKey` (boolean, optional): If `true`, attempts to detect an issue key from the branch name using the `keyRegex`.
    -   `keyRegex` (RegExp, optional): A regular expression used to detect issue keys in the branch name.
    -   `fallbackKey` (string, optional): A fallback issue key to use if none is detected in the branch name.
-   `format` (optional): An object containing properties related to commit message text formatting.
    -   `sentenceCase` (boolean, optional): If `false`, the first letter of the commit message will be in lowercase.
-   `scope`
    -   `mode` (string, optional): The commit scope mode, either `"monorepo"` or `"traditional"`. If `"monorepo"`, the scope will be the name of the package. If `"traditional"`, the scope will be the name of the directory containing the changed files. Defaults to `"monorepo"`.
    -   `srcDir` (string, optional): The source directory for the commit scope. Only used in `"traditional"` mode. Defaults to `"src"`.

### diff (optional)

This object contains properties related to the diff.

-   `maxSize` (number, optional): The maximum size (in bytes) of the diff that can be sent to the GPT model. If the diff exceeds this size, an error will be thrown.
-   `exclude` (string[], optional): An array of file paths to exclude from the diff.

## Usage

To use CodeBuddy, run the following command:

```
cb commit-all
```

this will do a git add for every detected scope, and generate a commit message, but for every available scope.

### Command-Line Arguments

CodeBuddy also supports the following command-line arguments:

-   `-r, --reason <reason>`: Specifies the reason for the changes, providing additional context for the commit message.
-   `-t, --type <type>`: Specifies the commit type, grounding the AI-generated message in the given type for better commit messages.
-   `-s, --scope <scope>`: Specifies the a specific scope, can help if order matters.

Some arguments coming soon:

-   `-b, --breaking`: Specifies that the changes are breaking changes.
-   `-i, --issue <issue>`: Specifies the issue key for the commit message (Note you can pull this from a commit message with config).

Please note that command-line arguments take precedence over the configuration file settings.

## Contributing

We welcome contributions! If you'd like to contribute, please follow these steps:

1. Fork the repository.
2. Create a new branch for your feature or bug fix.
3. Commit your changes and push them to your fork.
4. Create a pull request and describe the changes you made.

We will review your pull request and provide feedback.

## Changelog

Please refer to the [CHANGELOG.md](CHANGELOG.md) file for information about changes in each version.

## License

This project is licensed under the [ISC License](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to the CodeBuddy project will be documented in this file.

## 0.7.0

### Minor

-   🎉 (utils) Added new utility functions for handling changelog and package.json files ([no-key](#))
-   Added loadChangelog, loadPackageJson, writeChangelog, and writePackageJson functions in openai.ts
-   Imported fs module for file operations
-   Imported ScopeMode from types/index.js

-   🎉 (types) Added appName to the changelog in CodeBuddyConfig ([no-key](#))
-   The 'appName' field was added to the 'changelog' object in the 'CodeBuddyConfig' type.
-   This change affects the 'types' scope of the codebase.

-   🎉 (prompts) Automated versioning and changelog generation in prompts ([no-key](#))
-   Added functions to load and write package.json and changelog files
-   Implemented function to generate changelog from commit messages
-   Added functionality to increment version based on changes
-   Injected new changelog entries into existing changelog

## [0.6.0] - 23/07/2023

### Added

-   Added `maxTokens` as an option for the config file [INT-15].
-   Added `temperature` as an option for the config file [INT-15].
-   Added `topP` as an option for the config file [INT-15].
-   Added `frequencyPenalty` as an option for the config file [INT-15].
-   Added `presencePenalty` as an option for the config file [INT-15].
-   Added `stop` as an option for the config file [INT-15].

### Changed

-   Improved the error handling for commit message generation [INT-14].

### Fixed

-   Fixed a bug where the automatic issue key detection was not working [INT-16].

## [0.5.0] - 23/07/2023

### Added

-   OpenAI Function Calling to generate a consistently formatted commit message for the user.

### Changed

-   Updated the maxTokens to 900 up from 100 to allow for more complete commit message generation.

## [0.4.3] - 30/05/2023

### Fixed

-   Fixed a bug where the the diff check would fail if there were deleted files in the diff, due to a missing `--` in the `git.diff` command.

## [0.4.2] - 29/05/2023

### Fixed

-   Fixed a bug where the issue key was coming through as [undefined] instead of using the fallback key.

## [0.4.1] - 28/05/2023

### Changed

-   Updated README to include NPM installation instructions.
-   Updated README to include various badges.

## [0.4.0] - 27/05/2023

### Added

-   New `applyConfigTransform` and `applyConfigTransformAsync` functions to handle synchronous and asynchronous config transformations respectively.
-   New `applyScopeTrim` function for trimming specific strings from the commit scope.
-   Updated `getScopeMonorepo` in `git.ts` to accept an array of directories for better customization.
-   Added support for emoji in commit messages, enhancing their visual appeal.
-   Added an optional `reason` argument to the `determineCommitMessage` function in `src/utils/openai.ts` to provide additional context for the generated commit messages.
    -   This feature allows AI-generated commits to be coupled with a reason, providing business context where needed.
-   Added an optional `type` argument to the `determineCommitMessage` function in `src/utils/openai.ts` to provide a commit type for the generated commit messages.
    -   This feature allows AI-generated commits to be grounded in a specific type, improving the quality of the generated commit messages.

### Changed

-   Refactored `commitConfirmationPrompt` function in `src/prompts/index.ts` to use `applyConfigTransform` and `applyConfigTransformAsync` for applying configuration settings to the generated commit message. This change improves code readability and maintainability.
-   Updated the function signature and prompt generation in `determineCommitMessage` to include the reason when provided.

## [0.3.0] - 13/05/2023

### Added

-   New configuration options:
    -   `scope`: optional object under `commit` that allows for customization of the commit scope mode
        -   `mode`: controls the commit scope mode, either `"monorepo"` or `"traditional"`
        -   `srcDir`: controls the source directory for the commit scope (only used in `"traditional"` mode)
    -   `diff`: optional object for diff configuration
        -   `exclude`: optional array of strings to exclude from the diff (e.g. `["package-lock.json", "yarn.lock"]`)

### Changed

-   Refactored the `index.ts` to use `commander` for CLI argument parsing and built in help.

### Fixed

-   Fixed a bug where deleted files were incorrectly being sent through `git.add`

### Removed

-   Removed various unused utility functions from `utils/git.ts`

## [0.2.0] - 10/05/2023

### Added

-   Issue key detection from branch names
-   New configuration options:
    -   `useIssueKey`: allows issue key automation in commit messages
    -   `sentenceCase`: controls whether to use sentence case for commit messages

### Changed

-   Restructured `CodeBuddyConfig` to group related properties, improving organization and maintainability
-   Updated the `commitMessagePrompt` function in `src/prompts/index.ts` to handle new configuration options
-   Improved conciseness in commit message rules for both summary and description

## [0.1.0] - 08/05/2023

### Added

-   Initial release with basic functionality
-   Automatically generate commit messages based on the diff
    $$


src/utils/openai.ts
import {
    ChatCompletionRequestMessageRoleEnum,
    ChatCompletionResponseMessage,
    Configuration,
    OpenAIApi,
} from "openai";

import fs from "fs";
import path from "path";
import { CodeBuddyConfig, CommitMessageError, ConfigLoadError, ScopeMode } from "../types/index.js";

const DEFAULT_MAX_TOKENS = 200;
const DEFAULT_TEMPERATURE = 0.3;
const DEFAULT_TOP_P = 1;
const DEFAULT_FREQUENCY_PENALTY = 0.5;
const DEFAULT_PRESENCE_PENALTY = 0.5;

export async function loadConfig(): Promise<CodeBuddyConfig> {
    try {
        const config = await import(path.resolve(process.cwd(), "cb.config.js"));
        return config.default;
    } catch (error) {
        throw new ConfigLoadError("Error loading configuration file: " + error.message);
    }
}

export async function loadChangelog(): Promise<string> {
    try {
        const changelogPath = path.resolve(process.cwd(), "CHANGELOG.md");
        const changelog = await fs.promises.readFile(changelogPath, "utf8");
        return changelog;
    } catch (error) {
        throw new ConfigLoadError("Error loading changelog file: " + error.message);
    }
}

export async function loadPackageJson(): Promise<any> {
    try {
        const packageJsonPath = path.resolve(process.cwd(), "package.json");
        const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf8"));
        return packageJson;
    } catch (error) {
        throw new ConfigLoadError("Error loading package.json file: " + error.message);
    }
}

export async function writeChangelog(changelog: string): Promise<void> {
    try {
        const changelogPath = path.resolve(process.cwd(), "CHANGELOG.md");
        await fs.promises.writeFile(changelogPath, changelog, "utf8");
    } catch (error) {
        throw new ConfigLoadError("Error writing to changelog file: " + error.message);
    }
}

export async function writePackageJson(packageJson: any): Promise<void> {
    try {
        const packageJsonPath = path.resolve(process.cwd(), "package.json");
        await fs.promises.writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2), "utf8");
    } catch (error) {
        throw new ConfigLoadError("Error writing to package.json file: " + error.message);
    }
}

async function getOpenAI(): Promise<[OpenAIApi, CodeBuddyConfig]> {
    const config = await loadConfig();
    return [
        new OpenAIApi(
            new Configuration({
                apiKey: config.chatGPT.apiKey,
                organization: config.chatGPT.organization,
            })
        ),
        config,
    ];
}

const extractFunctionArgs = (response: ChatCompletionResponseMessage) => {
    const args = JSON.parse(response.function_call.arguments);
    return args;
};

/**
 * This function generates a commit message for a given diff, scope and type.
 *
 * @param diff The diff to generate a commit message for
 * @param scope The scope of the commit
 * @param type The type of the commit
 *
 * @returns The generated commit message
 *
 * @remarks
 * This function uses the OpenAI API to generate a commit message for a given diff, scope and type.
 * The type is optional, but if provided, will help GPT generate a more accurate commit message by grounding the commit message
 * in the type. If none is provided, GPT will generate a commit message and try to determine the type of the commit message.
 */
export async function determineCommitMessage(
    diff: string,
    scope: string,
    type?: string,
    reason?: string
) {
    const [openai, config] = await getOpenAI();
    const prompt = getCommitMessagePrompt(diff, scope, type, reason);

    try {
        const response = await openai.createChatCompletion({
            messages: [
                {
                    content: prompt,
                    role: ChatCompletionRequestMessageRoleEnum.User,
                },
            ],
            functions: [
                {
                    name: "structure-commit-message",
                    description:
                        "A function to structure a commit message based on changes and user context.",
                    parameters: {
                        type: "object",
                        properties: {
                            summary: {
                                type: "string",
                                description:
                                    "A concise summary of the changes to be committed, no longer than 100 words",
                            },
                            description: {
                                type: "string",
                                description:
                                    "A more detailed description of the changes to be committed as dot points, no longer than 200 words",
                            },
                            type: {
                                type: "string",
                                enum: [
                                    "feat",
                                    "fix",
                                    "docs",
                                    "style",
                                    "refactor",
                                    "test",
                                    "chore",
                                    "perf",
                                    "ci",
                                    "build",
                                    "revert",
                                ],
                            },
                        },
                        required: ["summary", "description"],
                    },
                },
            ],
            function_call: {
                name: "structure-commit-message",
            },
            model: config.chatGPT.model,
            max_tokens: config.chatGPT.maxTokens || DEFAULT_MAX_TOKENS,
            temperature: config.chatGPT.temperature || DEFAULT_TEMPERATURE,
            top_p: config.chatGPT.topP || DEFAULT_TOP_P,
            frequency_penalty: config.chatGPT.frequencyPenalty || DEFAULT_FREQUENCY_PENALTY,
            presence_penalty: config.chatGPT.presencePenalty || DEFAULT_PRESENCE_PENALTY,
        });

        if (!response.data.choices || response.data.choices.length === 0) {
            throw new CommitMessageError("GPT returned an empty response.");
        }

        const args = extractFunctionArgs(response.data.choices[0].message);

        return createMessage({
            scope,
            ...args,
        });
    } catch (error) {
        if (error instanceof CommitMessageError) {
            console.error("Commit message error:", error.message);
        } else {
            console.error("Error generating commit message:", error.message);
        }
        // Now, throw the error up the call stack so that it can be handled appropriately by the calling code
        throw error;
    }
}

const createMessage = ({
    scope,
    summary,
    description,
    type,
}: {
    scope: string;
    summary: string;
    description: string;
    type: string;
}) => {
    if (scope === ".") return `${type}: ${summary} \n\n${description}`;
    return `${type}(${scope}): ${summary} \n\n${description}`;
};

const commitTypeDescriptions = {
    docs: "You're updating documentation or code comments.",
    refactor: "You're making a code change that neither fixes a bug nor adds a feature.",
    fix: "You're fixing a bug.",
    feat: "You're adding a new feature.",
    style: "You're making changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc).",
    test: "You're adding missing tests or correcting existing tests.",
    chore: "You're making changes to the build process or auxiliary tools and libraries such as documentation generation.",
    perf: "You're making a code change that improves performance.",
    ci: "You're making changes to your CI configuration files and scripts.",
    build: "You're making changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm).",
    revert: "You're reverting a previous commit.",
};

const getCommitMessagePrompt = (diff: string, scope: string, type?: string, reason?: string) => {
    let commitTypes = `Commit Types:
- docs: When the changes affect files of type \`.md\` or pure code comments
- refactor: A code change that neither fixes a bug nor adds a feature
- fix: A bug fix
- feat: A new feature
- style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- test: Adding missing tests or correcting existing tests
- chore: Changes to the build process or auxiliary tools and libraries such as documentation generation
- perf: A code change that improves performance
- ci: Changes to your CI configuration files and scripts
- build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
- revert: Reverts a previous commit
`;

    let typeSpecificGuidelines = "";

    if (type) {
        commitTypes = "";
        typeSpecificGuidelines = `The commit is of type "${type}", so the summary should clearly indicate what changes were made under this type. ${commitTypeDescriptions[type]}. The description should provide a brief explanation of the changes, highlighting the main points and reasoning behind the changes made.`;
    }

    let reasonDescription = "";
    if (reason) {
        reasonDescription = `The changes were made because ${reason}. `;
    }

    return `Here are my changes:
\`${diff}\`

${type ? ` - The type of commit is "${type}". ` : " - I need to determine the type of commit. "}
${
    scope !== "."
        ? ` - The scope of the changes is "${scope}". `
        : " - This commit affects more than one scope. "
}

${reasonDescription}
${typeSpecificGuidelines}
${commitTypes}`;
};

const systemPromptMonorepo =
    'Generate a changelog from the provided commit messages, organizing them under the scope with suggested version increments (+x.y.z). Include explanations for the suggested increments and organize the commits into "Major", "Minor", and "Revisions" categories. Omit any \'chore\' commits. If a commit contains \'BREAKING\' (placed after the commit message and before the issue key), suggest a major version increment. Include emojis if present in the commit messages.\n\n' +
    "Given the following sample commit messages:\n\n" +
    [
        "feat(ui): 🎉 added usePrevious hook to the exports in inspace-ui [IN-889]\n\n- A new hook 'usePrevious' has been added to the exports in the hooks index file of inspace-ui package.",
        "refactor(search): 🧹 changed 'sector' to 'sectors' in filter constants [IN-889]\n\n- Modified the type of sectors from 'sector' to 'sectors' in the defaultFilterSchema and defaultFilterSchemaWorkspaces objects within the inspace-search package.",
        "refactor(api): 🧹 updated API endpoints BREAKING [IN-900]\n\n- The API endpoints were updated to improve performance and security. This change is not backward compatible.",
    ].join("\n\n") +
    "The expected output would be:\n\n" +
    `# Changelog

## ui +0.1.0

The version increment is suggested due to the addition of new features that do not break backward compatibility and enhance the functionality of the \`ui\` scope.

### Minor
- 🎉 Added \`usePrevious\` hook to the exports in inspace-ui ([IN-889](link-to-issue-IN-889))
  - A new hook 'usePrevious' has been added to the exports in the hooks index file of inspace-ui package.

## search +0.0.1

The version increment is suggested because the refactor enhances code quality and readability without adding new features or fixing bugs.

### Revisions
- 🧹 Changed 'sector' to 'sectors' in filter constants ([IN-889](link-to-issue-IN-889))
  - Modified the type of sectors from 'sector' to 'sectors' in the defaultFilterSchema and defaultFilterSchemaWorkspaces objects within the inspace-search package.

## api +1.0.0

The introduction of breaking changes to the API warrants a major version increment to signify the non-backward-compatible modifications.

### Major
- 🧹 Updated API endpoints (BREAKING CHANGE) ([IN-900](link-to-issue-IN-900))
  - The API endpoints were updated to improve performance and security. This change is not backward compatible.`;

const getSystemPromptSingleRepo = (appName: string) =>
    `For the app called ${appName}, generate a changelog from the provided commit messages, organizing them under the scope (if available) with suggested version increments (+x.y.z). Include explanations for the suggested increments and organize the commits into "Major", "Minor", and "Revisions" categories. Omit any \'chore\' commits. If a commit contains \'BREAKING\' (placed after the commit message and before the issue key), suggest a major version increment. Include emojis if present in the commit messages. \n\n` +
    "Given the following sample commit messages:\n\n" +
    [
        "feat: added changelog command and updated commit-all command in src/index.ts  [no-key]\n\n- Imported generateChangelog from ./prompts/index.js\n- Initialized commit-all command with version 0.2.0\n- Added options for reason and destination branch\n- Initialized changelog command with description and option for destination branch",
        "feat(utils): added getCommitLogs and getChangelog functions in utils  [no-key]\n\n- Added a new function 'getCommitLogs' in git.ts under utils.\n- Added another function 'getChangelog' in openai.ts under utils.",
        "feat(types): add changelog destination to CodeBuddyConfig type  [no-key]\n\n- Added a new `changelog` field to the `CodeBuddyConfig` type in `src/types/index.ts`.",
        "feat(prompts): added generateChangelog function and updated imports in prompts/index.ts  [no-key]\n\n- Added a new function generateChangelog to prompts/index.ts\n- Updated the import statement to include getCommitLogs from utils/git.js",
    ].join("\n\n") +
    "The expected output would be:\n\n" +
    `# Changelog

## +0.1.0

The version increment is suggested due to the addition of new features that do not break backward compatibility and enhance the functionality of the project.

### Minor
- 🎉 Added changelog command and updated commit-all command in src/index.ts ([no-key](#))
- Imported generateChangelog from ./prompts/index.js
- Initialized commit-all command with version 0.2.0
- Added options for reason and destination branch
- Initialized changelog command with description and option for destination branch

- 🎉 (utils) Added getCommitLogs and getChangelog functions in utils ([no-key](#))
- Added a new function 'getCommitLogs' in git.ts under utils.
- Added another function 'getChangelog' in openai.ts under utils.

- 🎉 (types) Add changelog destination to CodeBuddyConfig type ([no-key](#))
- Added a new \`changelog\` field to the \`CodeBuddyConfig\` type in \`src/types/index.ts\`.

- 🎉 (prompts) Added generateChangelog function and updated imports in prompts/index.ts ([no-key](#))
- Added a new function generateChangelog to prompts/index.ts
- Updated the import statement to include getCommitLogs from utils/git.js`;

export const getChangelog = async (commits: string) => {
    const [openai, config] = await getOpenAI();

    const userMessage = commits;

    // call openAi api
    const response = await openai.createChatCompletion({
        messages: [
            {
                content:
                    config.commit?.scope.mode === ScopeMode.Monorepo
                        ? systemPromptMonorepo
                        : getSystemPromptSingleRepo(config.changelog?.appName || "CodeBuddy"),
                role: ChatCompletionRequestMessageRoleEnum.System,
            },
            {
                content: userMessage,
                role: ChatCompletionRequestMessageRoleEnum.User,
            },
        ],
        model: config.chatGPT.model,
        max_tokens: config.chatGPT.maxTokens || DEFAULT_MAX_TOKENS,
        temperature: config.chatGPT.temperature || DEFAULT_TEMPERATURE,
        top_p: config.chatGPT.topP || DEFAULT_TOP_P,
        frequency_penalty: config.chatGPT.frequencyPenalty || DEFAULT_FREQUENCY_PENALTY,
        presence_penalty: config.chatGPT.presencePenalty || DEFAULT_PRESENCE_PENALTY,
    });

    if (!response.data.choices || response.data.choices.length === 0) {
        throw new Error("GPT returned an empty response.");
    }

    return response.data.choices[0].message;
};
